Section: A queueing model with 2 consecutive buffer centres

MichalisPanayides opened the pull request:
Queueing section directory tree:

    |-- queueing_model/
        |-- <subsections/>
            |-- <performance_measures/>
                |-- <blocking_time/>
                    |-- main.tex
                |-- <proportion_within_target/>
                    |-- main.tex
                |-- <waiting_time/>
                    |-- main.tex
            |-- <example/>
                |-- main.tex
        |-- main.tex

MichalisPanayides committed:
Add bibliography entry

MichalisPanayides committed:
Write details of the Markov-chain model

- Define the model
- Define state space
- Define transition matrix and state-mapping function

MichalisPanayides committed:
Add general Markov-chain picture

MichalisPanayides committed:
Update known.py

drvinceknight commented:
@11michalis11 the diff includes #5 so I'll wait till that's merged to review (just so it's easier to review).

MichalisPanayides committed:
Add .pdf file to GitHub repo

MichalisPanayides committed:
Create subsections/ subdirectory

MichalisPanayides committed:
Minor edits on figures

Reference figure of previous section and change the caption of
the general Markov chain figure

MichalisPanayides committed:
Add paragraph for transition rates

MichalisPanayides committed:
Apply minor edits

MichalisPanayides committed:
Write main.tex for performance measures subsection

MichalisPanayides committed:
Write main.tex for waiting time subsubsection

MichalisPanayides committed:
Update known.py

MichalisPanayides committed:
Rename directory

MichalisPanayides committed:
Fix typo

drvinceknight commented on sections/queueing_model/main.tex:
```diff
@@ -1,5 +1,115 @@
 
-\subsection{Example}
\ No newline at end of file
+The model consists of two types of individuals; class 1 and class 2.
+Class 1 individuals arrive instantly at waiting zone 1 and proceed to wait to
+receive their service. 
+Class 2 individuals arrive at waiting zone 2 and wait there until they are 
+allowed to move to waiting zone 1. 
```
Should we just do away with the word `Class`?

```suggestion
The model consists of two types of individuals.
Individuals of the first type arrive at waiting zone 1 and proceed to wait to
receive their service. 
Individuals of the second type arrive at waiting zone 2 and wait there until they are 
allowed to move to waiting zone 1. 
```

MichalisPanayides replied:
Any particular reason why we should? Is it because classes in queueing models are usually used to distinguish priority levels?

drvinceknight commented on sections/queueing_model/main.tex:
```diff
@@ -1,5 +1,115 @@
+receive their service. 
+Class 2 individuals arrive at waiting zone 2 and wait there until they are 
+allowed to move to waiting zone 1. 
+They are allowed to proceed to the next waiting zone only when the number of 
```
```suggestion
They are allowed to proceed only when the number of 
```

drvinceknight commented on sections/queueing_model/main.tex:
```diff
@@ -1,5 +1,115 @@
+Class 2 individuals arrive at waiting zone 2 and wait there until they are 
+allowed to move to waiting zone 1. 
+They are allowed to proceed to the next waiting zone only when the number of 
+individuals in the next system (in waiting zone 1 and in service) is less than a 
```
```suggestion
individuals in waiting zone 1 \textbf{and} in service is less than a 
```

drvinceknight commented on sections/queueing_model/main.tex:
```diff
@@ -1,5 +1,115 @@
+They are allowed to proceed to the next waiting zone only when the number of 
+individuals in the next system (in waiting zone 1 and in service) is less than a 
+pre-determined threshold \(T\).
+When the threshold is reached, all class 2 individuals that arrive will remain 
```
```suggestion
When the number of individuals is equal to or exceeds this threshold, all second type individuals that arrive will remain 
```

drvinceknight commented on sections/queueing_model/main.tex:
```diff
@@ -1,5 +1,115 @@
+individuals in the next system (in waiting zone 1 and in service) is less than a 
+pre-determined threshold \(T\).
+When the threshold is reached, all class 2 individuals that arrive will remain 
+\textit{``blocked''} in the buffer centre until the number of people in the 
```
```suggestion
\textit{``blocked''} in waiting zone 1 until the number of people in the 
```

drvinceknight commented on sections/queueing_model/main.tex:
```diff
@@ -1,5 +1,115 @@
+
+\begin{itemize}
+    \item \(\lambda_1\): The arrival rate of class 1 individuals
+    \item \(\lambda_2\): The arrival rate of class 2 individuals
```
```suggestion
    \item \(\lambda_i\): The arrival rate of individuals of type \(i\in\{1, 2\}\)
```

drvinceknight commented on sections/queueing_model/main.tex:
```diff
@@ -1,5 +1,115 @@
+    \item \(\lambda_1\): The arrival rate of class 1 individuals
+    \item \(\lambda_2\): The arrival rate of class 2 individuals
+    \item \(\mu\): The service rate for individuals receiving service in waiting 
+    zone 1
```
```suggestion
    \item \(\mu\): The service rate for individuals receiving service
```

drvinceknight commented on sections/queueing_model/main.tex:
```diff
@@ -1,5 +1,115 @@
+    \item \(\mu\): The service rate for individuals receiving service in waiting 
+    zone 1
+    \item \(C\): The number of servers
+    \item \(T\): The threshold where class 2 individuals will start being blocked 
```
```suggestion
    \item \(T\): The threshold at which individuals of the second type are blocked
```

drvinceknight commented on sections/queueing_model/main.tex:
```diff
@@ -1,5 +1,115 @@
+~\cite{kemeny1976markov}. 
+The way such models are formulated is through a directed graph where each vertex
+is consider a state of the system and each edge represents the transition rate
+between its two linking vertices.
```
```suggestion
Under the assumption that all rates (arrival and service) are Markovian the queuing system corresponds to a Markov chain~\cite{kemeny1976markov}.
```

drvinceknight commented on sections/queueing_model/main.tex:
```diff
@@ -1,5 +1,115 @@
+
+\begin{itemize}
+    \item \(u\) = number of individuals blocked
+    \item \(v\) = number of individuals either in waiting zone 1 or in the
```
```suggestion
    \item \(u\) is the number of individuals blocked
    \item \(v\) is the number of individuals either in waiting zone 1 or in the
```

drvinceknight commented on sections/queueing_model/main.tex:
```diff
@@ -1,5 +1,115 @@
+    service centre
+\end{itemize}
+
+
```
```suggestion
```

drvinceknight commented on sections/queueing_model/main.tex:
```diff
@@ -1,5 +1,115 @@
+The state space of the Markov chain \(S\) is defined as the union between
+\(S_1\) and \(S_2\), where \(S_1\) represents the set of states where the total 
+number of individuals is less than \(T\) and \(S_2\) the set of states where the
+total number of individuals is greater than \(T\). 
```
```suggestion
We denote the state space of the Markov chain as  \(S=S(T)\) which can be written as the disjoint union~\ref{eqn:definition_of_S_as_disjoint_union}.
```

drvinceknight commented on sections/queueing_model/main.tex:
```diff
@@ -1,5 +1,115 @@
+total number of individuals is greater than \(T\). 
+
+\begin{align}
+    S(T) =& S_1(T) \cup S_2(T) \text{ where:} \nonumber \\
```
```suggestion
    S(T) =& S_1(T) \cup S_2(T) \text{ where:} \label{eqn:definition_of_S_as_disjoint_union} \\
```

drvinceknight commented on sections/queueing_model/main.tex:
```diff
@@ -1,5 +1,115 @@
+    S_1(T) =& \left\{(0, v)\in\mathbb{N}_0^2 \; | \; v < T \right\} 
+    \label{eq:state_space} \\
+    S_2(T) =& \{(u, v)\in\mathbb{N}_0^2 \; | \; v \geq T \} \nonumber
+\end{align}
```
```suggestion
\end{align}

\(S_1\) represents the set of states where the total 
number of individuals is less than \(T\) and \(S_2\) the set of states where the
total number of individuals is greater or equal to \(T\). 
```

drvinceknight commented on sections/queueing_model/main.tex:
```diff
@@ -1,5 +1,115 @@
+    S_2(T) =& \{(u, v)\in\mathbb{N}_0^2 \; | \; v \geq T \} \nonumber
+\end{align}
+
+The transition matrix \(Q\) of the Markov chain consists of the transition rates
```
If we go with `\(Q\)` then we need to go with `\(Q_{ij}\)`.

Can you change all `q_{...}` to `Q_{...}`

Will catch a few here as I go.

MichalisPanayides replied:
Yeah don't know why I used `q_{i, j}` in the first place

drvinceknight commented on sections/queueing_model/main.tex:
```diff
@@ -1,5 +1,115 @@
+\end{align}
+
+The transition matrix \(Q\) of the Markov chain consists of the transition rates
+between the numerous states of the model. Every entry \( q_{i,j} = 
+q_{(u_i, v_i),(u_j, v_j)} \) represents the transition rate from state 
+\( i = (u_i, v_i) \) to state \( j = (u_j , v_j) \) for all 
+\( (u_i, v_i), (u_j, v_j) \in S \).
+The entries of \(Q\) can be calculated using the state-mapping function 
+described in \ref{eq:markov_transition_rate}: 
```
```suggestion
between the numerous states of the model. For a given ordering of \(S\), every entry \( Q_{ij} = 
Q_{(u_i, v_i),(u_j, v_j)} \) represents the transition rate from the \(i^{\text{th}}\) state 
\((u_i, v_i) \) to the \(j^{\text{th}}\) state \( (u_j , v_j) \) for all 
\( (u_i, v_i), (u_j, v_j) \in S \).
\(Q\) is given in~\ref{eq:markov_transition_rate}.
```

Note when you have just got two single letters as indicates you don't need to use a `,` so:

- `Q_{ij}` 👍  but `Q_{i,j}` 👎 
- `Q_{i + 1, j} :+1:

MichalisPanayides replied:
Had no idea there's convention for this. Good to know 👍

drvinceknight commented on sections/queueing_model/main.tex:
```diff
@@ -1,5 +1,115 @@
+described in \ref{eq:markov_transition_rate}: 
+
+\begin{equation} \label{eq:markov_transition_rate}
+    q_{i, j} = 
```
```suggestion
    Q_{i, j} = 
```

drvinceknight commented on sections/queueing_model/main.tex:
```diff
@@ -1,5 +1,115 @@
+\input{sections/queueing_model/general_markov_model/main.tex}
+
+
+In order to acquire an exact solution of the problem a slight adjustment needs to 
+be considered. 
```
You have not described the problem yet. Here you need to describe the steady state vector.

MichalisPanayides replied:
Yeah that makes sense. I properly define the steady state vector in `performance_measures/main.tex` but it makes more sense to have it here instead. Will do 👍

drvinceknight reviewed (changes requested):
Looking good so far @11michalis11

MichalisPanayides committed:
Apply suggestions from code review

Co-authored-by: Vince Knight <vincent.knight@gmail.com>

MichalisPanayides committed:
Apply more siggestions from review

- Markov chains intro
- State space definition
- Replace q_{i,j} with Q_{ij}
- Minor changes

MichalisPanayides committed:
Update known.py

MichalisPanayides committed:
Minor edits

MichalisPanayides committed:
Add blocking time algorithm

MichalisPanayides committed:
Add python code that generates tikz picture

In order for the code to work, ambulance_game needs
to be imported first.

MichalisPanayides committed:
Place all tikz pictures in imgs\ directory

MichalisPanayides committed:
Remove some un-needed comments

MichalisPanayides committed:
Update known.py

MichalisPanayides committed:
Add another markov chain example plot

MichalisPanayides committed:
Apply minor changes to some .tex files

MichalisPanayides committed:
Add proportion within target subsubsection

Add subsubsection that explains how the proportion of individuals
is calculated. There are several subparagraphs to that section that
explain additional components that may or may not be left in at the
end.

MichalisPanayides committed:
Update CI

MichalisPanayides committed:
Comment out section

MichalisPanayides committed:
Change bibliography and comment mathtools package

MichalisPanayides committed:
Use plain bibliography style

MichalisPanayides committed:
Add known words

MichalisPanayides committed:
Use mathtools package

MichalisPanayides committed:
Update CI and fix typo

Remove unwanted tex libraries

MichalisPanayides committed:
Include section that was commented out

MichalisPanayides committed:
Minor change on input

MichalisPanayides committed:
Comment most of proportion section out

MichalisPanayides committed:
Undo last commit

MichalisPanayides committed:
Commit .pdf as well

MichalisPanayides committed:
Merge pull request #7 from 11michalis11/fix-possible-issue-with-ci

Fix possible issue with CI

drvinceknight commented on sections/queueing_model/main.tex:
```diff
@@ -1,5 +1,105 @@
+In this section, a more in-depth explanation of the queueing model shown in 
+figure \ref{fig:diagram_of_queueing_system} will be given.
+This is a queuing model that consists of two waiting spaces, one for each type
+of patient.
```
```suggestion
of individual.
```

drvinceknight commented on sections/queueing_model/main.tex:
```diff
@@ -1,5 +1,105 @@
+\input{imgs/general_markov_model/main.tex}
+
+
+In order to acquire an exact solution of the problem a slight adjustment needs 
```
```suggestion
In order to consider this model numerically an adjustment needs 
```

drvinceknight commented on sections/queueing_model/main.tex:
```diff
@@ -1,5 +1,105 @@
+
+
+In order to acquire an exact solution of the problem a slight adjustment needs 
+to be considered. 
```
```suggestion
to be made. 
```

drvinceknight commented on sections/queueing_model/main.tex:
```diff
@@ -1,5 +1,105 @@
+to be considered. 
+The problem defined above assumes no upper boundary to the number of individuals 
+that can wait for service or for the ones that are blocked in the buffer centre. 
+Therefore, a different state space \( \tilde S \) needs to be constructed where 
+\( \tilde S \subseteq S \) and there is a maximum allowed number of people \(N\) 
+that can be in the system and a maximum allowed number of people \(M\) that can
```
```suggestion
Therefore, a different state space \( \tilde S \) needs to be constructed where 
\( \tilde S \subseteq S \) and there is a maximum allowed number of individuals \(N\) 
that can be in the system and a maximum allowed number of individuals \(M\) that can
```

drvinceknight commented on sections/queueing_model/main.tex:
```diff
@@ -1,5 +1,105 @@
+to be considered. 
+The problem defined above assumes no upper boundary to the number of individuals 
+that can wait for service or for the ones that are blocked in the buffer centre. 
+Therefore, a different state space \( \tilde S \) needs to be constructed where 
```
```suggestion
Therefore, a different state space \( \tilde S \) is constructed where 
```

drvinceknight commented on sections/queueing_model/subsections/performance_measures/main.tex:
```diff
@@ -0,0 +1,43 @@
+\subsection{Performance Measures}
+
+
+The transition matrix \( Q \) defined in \ref{eq:markov_transition_rate} can be 
```
```suggestion
The transition matrix \( Q \) defined in (\ref{eq:markov_transition_rate}) can be 
```

Equations always need to be referred to in brackets. Take a look through as I might have missed some of these.

drvinceknight commented on sections/queueing_model/subsections/performance_measures/main.tex:
```diff
@@ -0,0 +1,43 @@
+used to get the probability vector \( \pi \).
+The vector \( \pi \) is commonly used to study stochastic systems and it's main
+purpose is to keep track of the probability of being at any given state of 
+the system.
```
```suggestion
the system. \(\pi_i\) is the steady state probability of being in state \((u_i, v_i)\in\tilde S\) which is the \(i\)th state of \(tilde S\) for some ordering of \(\tilde S\).
```

drvinceknight reviewed (changes requested):
I've finished with the initial "simple" performance measures.

MichalisPanayides committed:
Apply suggestions from code review

Co-authored-by: Vince Knight <vincent.knight@gmail.com>

MichalisPanayides committed:
Change the way equations are cited

Include brackets around cited equations

MichalisPanayides committed:
Replace the word "class" with "type"

Class 1 and class 2 individulas are now referred to as
type 1 and type 2 individuals

drvinceknight commented on sections/queueing_model/subsections/performance_measures/waiting_time/main.tex:
```diff
@@ -0,0 +1,82 @@
+
+Consequently, the overall waiting time can be estimated by a linear combination 
+of \(W_1\) and \(W_2\). 
+Thus, the overall waiting time can calculated by the following equation where 
```
```suggestion
Thus, the overall waiting time can be calculated by the following equation where 
```

drvinceknight commented on sections/queueing_model/subsections/performance_measures/waiting_time/main.tex:
```diff
@@ -0,0 +1,82 @@
+\end{equation}
+
+The two coefficients represent the proportion of individuals of each type that 
+did not get lost and traversed through the model.
```
```suggestion
are not lost to the system.
```

drvinceknight commented on sections/queueing_model/subsections/performance_measures/waiting_time/main.tex:
```diff
@@ -0,0 +1,82 @@
+individual is not lost in the system is given by:
+
+\begin{equation*}
+    P(L'_1) = \sum_{(u,v) \, \in S_A^{(1)}} \pi(u,v) \hspace{2cm}
+    P(L'_2) = \sum_{(u,v) \, \in S_A^{(2)}} \pi(u,v)
+\end{equation*}
+ 
+Thus, by using these values as the coefficient of equation 
+(\ref{eq:overall_waiting_time_coeff}) the resultant equation can be used to get 
+the overall waiting time. 
+
+\begin{equation}\label{eq:overall_waiting_time}
+    W = \frac{\lambda_1 P(L'_1)}{\lambda_2 P(L'_2) + \lambda_1 P(L'_1)} W^{(1)} + 
+    \frac{\lambda_2 P(L'_2)}{\lambda_2 P(L'_2) + \lambda_1 P(L'_1)} W^{(2)}
```
It took me a little while to realise that `L'_i` was not a parameter. Change the notation to be:

```
P_{L_1} 
```

and 

```
P_{L_2}
```

MichalisPanayides replied:
@drvinceknight 

I'm using `P(L'_i)` to define the probability that an individual of type `i` is NOT lost (so basically the probability of accepting an individual of type `i`). I did not use `P(A_i)` because I'm using the letter `A` for `S_A` (the set of accepting states).

I understand that what I have now is kind of ugly-looking but I think the solution would be to either explain better that `L'_i` is the negation of `L_i` or just use `P(A_i)`.

I don't fully understand why you said I should use `P_{L_i}`. I was literally talking about the: 
`Probability that a type i individual is not lost` which I believe translates to `P(L'_i)`

drvinceknight replied:
`P(X)` makes it look like this is value parametrised by some parameter `X`. I'm fine with `P_{L'_i}`.

MichalisPanayides replied:
Great 👍. Will use that then.

drvinceknight commented on sections/queueing_model/subsections/performance_measures/waiting_time/main.tex:
```diff
@@ -0,0 +1,82 @@
+
+Waiting time is the amount of time that individuals from either type wait in 
+waiting zone 1 so that they can receive their service. 
+For a given set of parameters there are three different performance measures 
+around the mean waiting time that can be calculated; the mean waiting time of
+type 1 individuals, the mean waiting time of type 2 individuals and the 
+overall mean waiting time. 
```
Let us reorganise this section (and all the extra performance measure sections):

```suggestion
For a given set of parameters there are three different performance measures 
around the mean waiting time that can be calculated; the mean waiting time of
type 1 individuals:

\begin{equation}
% TODO Write expression here.
\end{equation}

The mean waiting time of type 2 individuals

\begin{equation}
% TODO Write expression here.
\end{equation}
 
The 
overall mean waiting time:

\begin{equation}
% TODO Write expression here.
\end{equation}
 
Where % TODO Write down what each expression/parameter is (if necessary).
```

Then move how you obtained this to an appendix.

drvinceknight replied:
We then want to include here, some figures with a comparison with simulated and calculated values. 

This does require an introduction of the simulation model but that can be short again.

```
Appendix A gives more details on obtaining equations .... 
Figure ... shows a comparison of the calculated values (using equations ...) and simulated using a discrete event simulation (details of the discrete event simulation model are given in Appendix B).
```

That figure will be 3 violin plots I believe.

MichalisPanayides replied:
The 3 violin plots will be for: `type 1, type 2, overall` performance measures over multiple runs of the the DES model. On the same plot there will also be the value of the performance measure for the Markov model. Is that right?

drvinceknight replied:
:+1:

drvinceknight reviewed (changes requested):
I've gone through the waiting time section.

drvinceknight commented on sections/queueing_model/subsections/performance_measures/blocking_time/main.tex:
```diff
@@ -0,0 +1,225 @@
+    S_b = \{(u,v) \in S \; | \; u > 0\}
+\end{equation}
+ 
+In order to not consider individuals that will be lost to the system, the set of 
+accepting states needs to be taken into consideration. 
```
```suggestion
```

drvinceknight commented on sections/queueing_model/subsections/performance_measures/blocking_time/main.tex:
```diff
@@ -0,0 +1,225 @@
+\end{equation*}
+
+The mean sojourn time for each state is given by the inverse of the out-flow of
+that state.
```
Cite Stewart here.

drvinceknight commented on sections/queueing_model/subsections/performance_measures/blocking_time/main.tex:
```diff
@@ -0,0 +1,225 @@
+    \item be because of a service being completed: we will denote the probability 
+    of this happening by \(p_s(u,v)\). 
+    \item be because of an arrival of an individual of type 1: denoting such 
+    probability by \(p_o(u,v)\).
```
Can we not use `p_a(u, v)` ?

MichalisPanayides replied:
Yeah, we can. Will change that now.

drvinceknight commented on sections/queueing_model/subsections/performance_measures/blocking_time/main.tex:
```diff
@@ -0,0 +1,225 @@
+            \vdots & \vdots & \vdots & \ddots & \vdots & \vdots & \vdots & 
+            \vdots & \vdots & \ddots & \vdots & \vdots \\ 
+            0 & 0 & 0 & \dots & 1 & -1 & 0 & 0 & 0 & \dots & 0 & 0 \\ %(1,N)
+            p_s & 0 & 0 & \dots & 0 & 0 & -1 & p_o & 0 & \dots & 0 & 0 \\ %(2,T)
```
IS there supposed to be a `p_s` in the first column of this row?

MichalisPanayides replied:
Yep. That is the probability that a service occurs on state `(2, T)`. Which means that the next state will be `(1, T)`. The columns (and rows) of matrix `Z` correspond to these states (in that order)
```
(1, T), (1, T+1), (1, T+2), ... , (1, N), (2, T), (2, T+1), (2, T+2), ... , (2, N), ... (M, T), (M, T+1), (M, T+2), ... , (M, N),
```

drvinceknight commented on sections/queueing_model/subsections/performance_measures/blocking_time/main.tex:
```diff
@@ -0,0 +1,225 @@
+            0 & 0 & 0 & \dots & 0 & 0 & p_s & -1 & p_o & \dots & 0 & 0 \\ %(2,T+1)
+            \vdots & \vdots & \vdots & \ddots & \vdots & \vdots & \vdots & 
+            \vdots & \vdots & \ddots & \vdots & \vdots \\ 
+            0 & 0 & 0 & \dots & 0 & 0 & 0 & 0 & 0 & \dots & 1 & -1 \\ %(M,N)
```
Can we write a general form for `Z_{ij}` here please?

MichalisPanayides replied:
@drvinceknight So remove the matrix completely and replace it with the general form `Z_{ij}` or keep both?

drvinceknight replied:
Keep both.

drvinceknight commented on sections/queueing_model/subsections/performance_measures/blocking_time/main.tex:
```diff
@@ -0,0 +1,225 @@
+the second type.  
+That is because individuals of the first type cannot be blocked. 
+Thus, one only needs to consider the pathway of type 2 individuals to get the 
+mean blocking time of the system. 
```
After this paragraph we need to write the "formula" and by that I mean: the linear system `b(u,v)=...`. (Equation `eq:general_blocking_time_at_each_state`) and that it can be solved by considering the linear system `Zx=y` where `Z_{ij} =...`, `x_{i}=...` and `y_j = ...`.

Then we move/replicate everything else in an appendix and include a similar numerical example.

drvinceknight reviewed (changes requested):
Same overall suggestion about the blocking time.

For the proportion within target, let's do the same overall change as well:

- Include what is essentially on the last page
- Move everything else to appendix (or sequence of appendix)
- Include small numeric example as a Figure.

This all will have the effect of "getting to the point" faster.

For your thesis, each to these things could be chapters in their own rights too they might still stay as an appendix there.

MichalisPanayides committed:
Apply suggestions to waiting time section from code review

Co-authored-by: Vince Knight <vincent.knight@gmail.com>

MichalisPanayides committed:
Restructure waiting time formula section

The section is now structured in the
following way:

- Formula 1
- Formula 2
- Formula 3
- Defining necessary parameters

MichalisPanayides committed:
Minor change in notation on waiting time formula

MichalisPanayides committed:
Restructure blocking time formula

Remove most of text that focuses on the "how" part of the formula.
What's left now is just the formula and everything that's necessary to
use it.

MichalisPanayides committed:
Add reference for sojourn time

MichalisPanayides committed:
Add an appendix

Currently the appendix has some information of how to use the blocking
time formula. Soon to be added are information on the waiting time
formula (possibly the recursive method) and information on the
proportion within target formula.

MichalisPanayides committed:
Restructure proportion within time target formula

Remove most parts of this section and move them to appendix

MichalisPanayides committed:
Add python code to generate violinplots

Added 3 python files that will generate violinplots for
the waiting time, blocking time and proportion within target
using the simulation model. These are then compared in the
same plot with the analytical outcomes using the Markov
chains method.

drvinceknight commented on sections/queueing_model/subsections/performance_measures/blocking_time/main.tex:
```diff
@@ -0,0 +1,125 @@
+            b(1,N) \\
+            b(2,T) \\
+            b(2,T+1) \\
+            \vdots \\
+            b(M,N) \\
+        \end{pmatrix}, 
+        y= 
+        \begin{pmatrix}
+            -c(1,T) \\
+            -c(1,T+1) \\
+            -c(1,T+2) \\
+            \vdots \\
+            -c(1,N) \\
+            -c(2,T) \\
+            -c(2,T+1) \\
+            \vdots \\
+            -c(M,N) \\
+        \end{pmatrix}
+    \)}
+\end{equation}
+
+Matrix \(Z\) can be alternatively defined using \(Z_{ij}\) defined in equation 
+(\ref{eq:general_mapping_function_of_blocking_matrix}) where \(i\) and \(j\) 
+represent states \((u_i, v_i), (u_j, v_j) \in S_b\).
+
+\begin{equation}\label{eq:general_mapping_function_of_blocking_matrix}
+    Z_{ij} = 
+    \begin{cases}
+        p_a, & \textbf{if } j = i + 1 \textbf{ and } v_i \neq N \\
+        \\
+        p_s, & \textbf{if } j = i - 1 \textbf{ and } v_i \neq N, v_i \neq T \\
+             & \textbf{or } j = i - N + T \textbf{ and } u_i \geq 2,\,v_i = T \\
+        \\
+        1, & \textbf{if } j = i - 1 \textbf{ and } v_i = N \\
+        \\
+        -1, & \textbf{if } i = j \\
+        \\
+        0, & \textbf{otherwise} \\
+    \end{cases}
+\end{equation}
```
```suggestion
The system of equations produced by (\ref{eq:general_blocking_time_at_each_state}) can be solved by considering the linear system \(Zx=y\). Assuming \(i\) and \(j\) 
represent states \((u_i, v_i), (u_j, v_j) \in S_b\) then \(Z_{ij}\) is given by:

\begin{equation}\label{eq:general_mapping_function_of_blocking_matrix}
    Z_{ij} = 
    \begin{cases}
        p_a, & \textbf{if } j = i + 1 \textbf{ and } v_i \neq N \\
        \\
        p_s, & \textbf{if } j = i - 1 \textbf{ and } v_i \neq N, v_i \neq T \\
             & \textbf{or } j = i - N + T \textbf{ and } u_i \geq 2,\,v_i = T \\
        \\
        1, & \textbf{if } j = i - 1 \textbf{ and } v_i = N \\
        \\
        -1, & \textbf{if } i = j \\
        \\
        0, & \textbf{otherwise} \\
    \end{cases}
\end{equation}

Equation~(\ref{eq:general_algebaric_approach_blocking_time}) shows this.

\begin{equation}\label{eq:general_algebaric_approach_blocking_time}
    \scalebox{0.73}{\(
        Z = 
        \begin{pmatrix}
            -1 & p_a & 0 & \dots & 0 & 0 & 0 & 0 & 0 & \dots & 0 & 0 \\ %(1,T)
            p_s & -1 & p_a & \dots & 0 & 0 & 0 & 0 & 0 & \dots & 0 & 0 \\ %(1,T+1)
            0 & p_s & -1 & \dots & 0 & 0 & 0 & 0 & 0 & \dots & 0 & 0 \\ %(1,T+2)
            \vdots & \vdots & \vdots & \ddots & \vdots & \vdots & \vdots & 
            \vdots & \vdots & \ddots & \vdots & \vdots \\ 
            0 & 0 & 0 & \dots & 1 & -1 & 0 & 0 & 0 & \dots & 0 & 0 \\ %(1,N)
            p_s & 0 & 0 & \dots & 0 & 0 & -1 & p_a & 0 & \dots & 0 & 0 \\ %(2,T)
            0 & 0 & 0 & \dots & 0 & 0 & p_s & -1 & p_a & \dots & 0 & 0 \\ %(2,T+1)
            \vdots & \vdots & \vdots & \ddots & \vdots & \vdots & \vdots & 
            \vdots & \vdots & \ddots & \vdots & \vdots \\ 
            0 & 0 & 0 & \dots & 0 & 0 & 0 & 0 & 0 & \dots & 1 & -1 \\ %(M,N)
        \end{pmatrix},
        x = 
        \begin{pmatrix}
            b(1,T) \\
            b(1,T+1) \\
            b(1,T+2) \\
            \vdots \\
            b(1,N) \\
            b(2,T) \\
            b(2,T+1) \\
            \vdots \\
            b(M,N) \\
        \end{pmatrix}, 
        y= 
        \begin{pmatrix}
            -c(1,T) \\
            -c(1,T+1) \\
            -c(1,T+2) \\
            \vdots \\
            -c(1,N) \\
            -c(2,T) \\
            -c(2,T+1) \\
            \vdots \\
            -c(M,N) \\
        \end{pmatrix}
    \)}
\end{equation}


```

MichalisPanayides committed:
Add comparisons between Markov and DES

Include violinplots of waiting time, blocking time and proportion
within target time. The violinplots show the differences between
Markov values and simulated values.

MichalisPanayides committed:
Add new plots to compare Markov and DES

The comparisons are now made over different values of
lambda_2, for all performance measures for all possible
types of individuals.

MichalisPanayides committed:
Add a todo to add description for DES

MichalisPanayides committed:
Fix latex compile issue with pdfs (#8)

* Remove scalebox before \includegraphics

* Add updated .pdf

* Chnage the way pdf compiles in CI

tasks.py are updated so that latexmk runs with the xelatex parameter

* Update CI to install additional libraries

Specifically: texlive-xetex, texlive-science

* Restructure imgs/ directory

MichalisPanayides committed:
Add new violinplots

Include violin plots of performance measures over
different values of lambda_2 for type 1, type 2 and
overall individuals

MichalisPanayides committed:
Apply changes from code review

MichalisPanayides committed:
Restructure queueing model section

Remove performance measures subsection and change "waiting time",
"blocking time" and "proportion of inds" from subsubsections into
subsections. Also add example subsubsections

MichalisPanayides committed:
Write appendix chapter for DES model

MichalisPanayides committed:
Edit minor things

MichalisPanayides committed:
Add recursive waiting time formula to appendix

MichalisPanayides committed:
Minor change

drvinceknight commented on sections/queueing_model/waiting_time/main.tex:
```diff
@@ -0,0 +1,104 @@
+(\ref{eq:closed_form_waiting_type_2}) and (\ref{eq:overall_waiting_time})
+originate from. 
+
+\subsubsection{Example}
```
```suggestion
```

drvinceknight commented on sections/queueing_model/waiting_time/main.tex:
```diff
@@ -0,0 +1,104 @@
+waiting time using discrete event simulation over a range of values of 
+\(\lambda_2\) (details of the discrete event simulation model are given in 
+appendix~\ref{sec:appendix_des}).
+The simulation was ran 100 times and the recorded mean waiting time at each 
+iteration is used to populate the violin plots.
+
+\begin{figure}[H]
+    \centering
+    \includegraphics[width=.8\textwidth]{imgs/waiting_time_comparison/waiting_1.pdf}
+    \caption{
+        Comparison of mean waiting time for type 1 individuals between values 
+        obtained from the Markov chain formulas and values obtained from 
+        simulation.
+    }
+    \label{fig:markov_vs_des_waiting_time_comparison_1}
+\end{figure}
```
Let us add 2 other versions of these.

- [ ] Include a table with 3 columns. The rows correspond to the parameters of the system (choose 3 relatively different systems).
- [ ] Plot the waiting time comparisons as 3 separate pdfs. This will lead to 3 subfigures for each figure. (3 plots next to each other)
- [ ] Include a bit more of a discussion in the captions (confirming obvious general trends as well as the fact that they match).

drvinceknight commented on sections/queueing_model/blocking_time/main.tex:
```diff
@@ -0,0 +1,148 @@
+values of the blocking time extracted from discrete event simulation
+(appendix~\ref{sec:appendix_des}).
+The blocking time is calculated using both methods for a range of values of
+\(\lambda_2\) 
```
You have not mentioned the 100 times repetition here?

MichalisPanayides replied:
Indeed I have not. Must have missed it

drvinceknight commented on sections/queueing_model/blocking_time/main.tex:
```diff
@@ -0,0 +1,148 @@
+(appendix~\ref{sec:appendix_des}).
+The blocking time is calculated using both methods for a range of values of
+\(\lambda_2\) 
+
+\begin{figure}[H]
+    \centering
+    \includegraphics[width=.8\textwidth]{imgs/blocking_time_comparison/blocking.pdf}
+    \caption{
+        Comparison of mean blocking time between values obtained from the 
+        Markov chain formulas and values obtained from simulation.
+    }
+    \label{fig:markov_vs_des_blocking_time_comparison}
+\end{figure}
```
Let us add 2 other versions of these.

- [ ] Include a table with 3 columns. The rows correspond to the parameters of the system (choose 3 relatively different systems). You might not need this and can just refer to the table in the waiting time section.
- [ ] Plot the blocking time comparisons as 3 separate pdfs.
- [ ] Include a bit more of a discussion in the captions (confirming obvious general trends as well as the fact that they match).

drvinceknight commented on sections/queueing_model/blocking_time/main.tex:
```diff
@@ -0,0 +1,148 @@
+\begin{equation}\label{eq:general_mapping_function_of_blocking_matrix}
+    Z_{ij} = 
+    \begin{cases}
+        p_a, & \textbf{if } j = i + 1 \textbf{ and } v_i \neq N \\
+        \\
+        p_s, & \textbf{if } j = i - 1 \textbf{ and } v_i \neq N, v_i \neq T \\
+             & \textbf{or } j = i - N + T \textbf{ and } u_i \geq 2,\,v_i = T \\
+        \\
+        1, & \textbf{if } j = i - 1 \textbf{ and } v_i = N \\
+        \\
+        -1, & \textbf{if } i = j \\
+        \\
+        0, & \textbf{otherwise} \\
```
```suggestion
        p_a, & \textbf{if } j = i + 1 \textbf{ and } v_i \neq N \\
        p_s, & \textbf{if } j = i - 1 \textbf{ and } v_i \neq N, v_i \neq T \\
             & \textbf{or } j = i - N + T \textbf{ and } u_i \geq 2,\,v_i = T \\
        1, & \textbf{if } j = i - 1 \textbf{ and } v_i = N \\
        -1, & \textbf{if } i = j \\
        0, & \textbf{otherwise} \\
```

drvinceknight commented on sections/queueing_model/proportion_within_target/main.tex:
```diff
@@ -0,0 +1,141 @@
+The simulation was ran 100 times and the recorded proportions at each iteration 
+is used to populate the violin plots.
+
+\begin{figure}[H]
+    \centering
+    \includegraphics[width=.8\textwidth]{imgs/proportion_within_target_comparison/proportion_1.pdf}
+    \caption{
+        Comparison of proportion within target time for type 1 individuals 
+        between values obtained from the Markov chain formulas and values 
+        obtained from simulation. 
+    }
+    \label{fig:markov_vs_des_proportion_comparison_1}
+\end{figure}
+
+\begin{figure}[H]
+    \centering
+    \includegraphics[width=.8\textwidth]{imgs/proportion_within_target_comparison/proportion_2.pdf}
+    \caption{
+        Comparison of proportion within target time for type 2 individuals 
+        between values obtained from the Markov chain formulas and values 
+        obtained from simulation. 
+    }
+    \label{fig:markov_vs_des_proportion_comparison_2}
+\end{figure}
+
+\begin{figure}[H]
+    \centering
+    \includegraphics[width=.8\textwidth]{imgs/proportion_within_target_comparison/proportion_overall.pdf}
+    \caption{
+        Comparison of proportion within target time for both type of 
+        individuals between values obtained from the Markov chain formulas 
```
Let us add 2 other versions of these.

- [ ] Include a table with 3 columns. The rows correspond to the parameters of the system (choose 3 relatively different systems). You might not need this and can just refer to the table in the waiting time section.
- [ ] Plot the proportion comparisons as 3 separate pdfs. This will lead to 3 subfigures for each figure. (3 plots next to each other)
- [ ] Include a bit more of a discussion in the captions (confirming obvious general trends as well as the fact that they match).

drvinceknight commented on sections/queueing_model/proportion_within_target/main.tex:
```diff
@@ -0,0 +1,141 @@
+(\ref{eq:proportion_within_target_type_1}) - 
+(\ref{eq:proportion_within_target_type_2_from_state}).
+
+\subsubsection{Example}
```
```suggestion
```

drvinceknight commented on sections/queueing_model/proportion_within_target/main.tex:
```diff
@@ -0,0 +1,141 @@
+    \begin{cases}
+        1 - \sum_{i=0}^{v-1} \frac{1}{i!} e^{-\mu t} (\mu t)^i, 
+            & \textbf{if } C = 1 \\
+            & \textbf{and } v>1 \\
+        & \\
+        1 - (\mu C)^{v-C} \mu  
+            \sum_{k=1}^{\mid \vec{r} \mid} \sum_{l=1}^{r_k}
+            \frac{\Psi_{k,l}(-\lambda_k)t^{r_k - l} 
+            e^{-\lambda_k t}}{(r_k - l)! (l - 1)!},
+            & \textbf{if } C > 1 \\
+        \textbf{where } \vec{r}=(v - C, 1) \textbf{ and } 
+            \vec{\lambda}=(C \mu, \mu) & \textbf{and } v > C \\
+            \hspace{1.15cm} \lambda_0 = 0, r_0 = 1 \\
+        & \\
+        1 - e^{-\mu t},  & \textbf{if } v \leq C
+    \end{cases}
+\end{equation}
+
+
+\begin{equation}\label{eq:proportion_within_target_type_2_from_state}
+    P(X_{(u,v)}^{(2)} < t) = 
+    \begin{cases}
+        1 - \sum_{i=0}^{\min(v,T)-1} \frac{1}{i!} e^{-\mu t} (\mu t)^i,  
+            & \textbf{if } C = 1 \\ 
+            & \textbf{and } v, T > 1 \\
+            & \\
+        1 - (\mu C) ^ {\min(v,T) - C} \mu  
+            &  \\
+        \qquad \times \sum_{k=1}^{\mid \vec{r} \mid} \sum_{l=1}^{r_k} 
+        \frac{\Psi_{k,l}(-\lambda_k)t^{r_k - l} 
+        e^{-\lambda_k t}}{(r_k - l)! (l - 1)!}, 
+            & \textbf{if } C > 1 \\
+        \textbf{where } \vec{r}=(\min(v, T) - C, 1) 
+            & \textbf{and } v, T  > C\\
+        \hspace{1.15cm} \vec{\lambda}=(C \mu, \mu) \\
+        \hspace{1.15cm} \lambda_0 = 0, r_0 = 1 \\
+            & \\
+        1 - e^{-\mu t}, & \textbf{if } v \leq C \\ 
+            & \textbf{or } T \leq C \\
+    \end{cases}
```
I started making LaTeX suggestions but it got messy:

- Remove the extra blank lines you add.
- Move the `\textbf{where} ...` statement to outside of the equation.

MichalisPanayides replied:
Will do.

drvinceknight commented on sections/queueing_model/proportion_within_target/main.tex:
```diff
@@ -0,0 +1,141 @@
+The overall proportion individuals within a time target (where \(P_{L'_1}\) and 
+\(P_{L'_1}\) are defined in (\ref{eq:proportion_of_accepting_individuals})):
+
+\begin{align}
+    P(X < t) &= \frac{\lambda_1 P_{L'_1}}{\lambda_2 P_{L'_2}+\lambda_1 P_{L'_1}} 
+    P(X^{(1)} < t) \nonumber \\
+    &+ \frac{\lambda_2 P_{L'_2}}{\lambda_2 P_{L'_2} + \lambda_1 
+    P_{L'_1}} P(X^{(2)} < t) \label{eq:overall_proportion_within_target}
+\end{align}
```
```suggestion
\begin{equation}
    P(X < t) &= \frac{\lambda_1 P_{L'_1}}{\lambda_2 P_{L'_2}+\lambda_1 P_{L'_1}} 
    P(X^{(1)} < t) + \frac{\lambda_2 P_{L'_2}}{\lambda_2 P_{L'_2} + \lambda_1 
    P_{L'_1}} P(X^{(2)} < t) \label{eq:overall_proportion_within_target}
\end{equation}
```

It's not necessary to put these across 2 lines.

drvinceknight replied:
> It's not necessary to put these across 2 lines.

If you do though the `+` needs to be offset.

drvinceknight commented on sections/appendix/descrete_event_simulation/main.tex:
```diff
@@ -0,0 +1,22 @@
+The queueing model was built in python using the Ciw 
+library~\cite{ciwpython}.
+
+% TODO: Include the queueing model diagram here?
+The constructed model simulates a queueing system with two waiting spaces and 
+two types of individuals. 
```
No just refer to the Figure.

drvinceknight commented on sections/appendix/descrete_event_simulation/main.tex:
```diff
@@ -0,0 +1,22 @@
+individuals moving from waiting zone 2 into waiting zone 1 may remain blocked 
+if there are more than \(T\) individuals in waiting zone 1.
+
+The same performance measures described in sections 
+\ref{sec:waiting_time}, \ref{sec:blocking_time} and 
+\ref{sec:proportion_within_target} can also be calculated using the DES model.
+The simulation can be ran a number of times to eliminate stochasticity and the
+outcomes of the two methods can be directly comparable. 
```
I believe this is incomplete?

drvinceknight commented on sections/appendix/descrete_event_simulation/main.tex:
```diff
@@ -0,0 +1,22 @@
+\ref{sec:proportion_within_target} can also be calculated using the DES model.
+The simulation can be ran a number of times to eliminate stochasticity and the
+outcomes of the two methods can be directly comparable. 
+
```
```suggestion
% TODO Add more documentation.
```

Imagine that this was the documentation for the simulation code. Include a code snippet to use it. Demonstrate the API etc... (We can discuss this).

drvinceknight commented on sections/appendix/blocking_time_formula/main.tex:
```diff
@@ -0,0 +1,227 @@
+
+    \begin{equation}
+    Z_{ij} = 
+    \begin{cases}
+        p_a, & \textbf{if } j = i + 1 \textbf{ and } v_i \neq N \\
+        \\
+        p_s, & \textbf{if } j = i - 1 \textbf{ and } v_i \neq N, v_i \neq T \\
+            & \textbf{or } j = i - N + T \textbf{ and } u_i \geq 2,\,v_i = T \\
+        \\
+        1, & \textbf{if } j = i - 1 \textbf{ and } v_i = N \\
+        \\
+        -1, & \textbf{if } i = j \\
+        \\
+        0, & \textbf{otherwise} \\
+    \end{cases}
+    \tag{\ref{eq:general_mapping_function_of_blocking_matrix} revisited}
```
```suggestion
    \begin{cases}
        p_a, & \textbf{if } j = i + 1 \textbf{ and } v_i \neq N \\
        p_s, & \textbf{if } j = i - 1 \textbf{ and } v_i \neq N, v_i \neq T \\
            & \textbf{or } j = i - N + T \textbf{ and } u_i \geq 2,\,v_i = T \\
        1, & \textbf{if } j = i - 1 \textbf{ and } v_i = N \\
        -1, & \textbf{if } i = j \\
        0, & \textbf{otherwise} \\
    \end{cases}
    \tag{\ref{eq:general_mapping_function_of_blocking_matrix} revisited}
```

drvinceknight reviewed (changes requested):
Most of my requests are stylistic but there are some more numeric experiments I'd like to see.

MichalisPanayides committed:
Apply minor changes from review
